### ISSUE TITLE
Adding microlensing propagation effects

### ISSUE THREAD

**Issue by jpierel14** [sncosmo/sncosmo#230]
Just opening this to discuss adding some features that I developed for my time delays software (https://arxiv.org/abs/1902.01260). The easiest is a microlensing propagation effect, which would entail adding optional phase dependence to existing propagation effect functions. I've also enabling light curve fitting inside of SNCosmo with the parameterized light curve model from Bazin et al. 2009, which could be added as a fitting option as well.

**Comment by rbiswas4** [comment]
The microlensing propagation effect could be quite interesting! You seem to be suggesting adding something along the lines of a propagation effect. Is the idea that these should use a random seed per SN to obtain different microlensing for different SN, or are you suggesting adding parameters to describe these?

**Comment by jpierel14** [comment]
Yes I've already made a microlensing propagation effect that works within the SNCosmo framework, but requires a small change that the existing (dust) effects have a phase argument that isn't used, so that the _flux function in the Model class can have a phase factor because the microlensing effect needs to be time-dependent. 

What I have already is just an effect defined by a curve in time+flux for achromatic, and time+wavelength+flux for chromatic microlensing. Then a user could just define their own microlensing curve for either simulation or fitting purposes. 

If you want to be able to generate these inside of SNCosmo, you could either define them using random splines (one option I have) or by actually convolving a SN expanding photosphere with a microcaustic generated by microlens. Actually simulating these curves definitely adds a degree of complication in terms of including it in SNCosmo though, the simple first thing to do is just have the effect so that a user can define a curve and include it in a model.

I toyed around with adding parameters to fit or generate these curves, but couldn't figure out a reasonable way of doing this because the effect is so stochastic. Open to ideas though!
